### PR TITLE
db: add tests to external services

### PR DESCRIPTION
**Pull request by unknwon**
Add tests for external services in DB layer.

Please review by commit. Inline comments are added for necessary elaboration.

Part of #10067

### CODE REVIEW

File: cmd/frontend/db/external_services.go [hunk 1]
```diff
@@ -62,11 +62,11 @@ type ExternalServicesListOptions struct {
 func (o ExternalServicesListOptions) sqlConditions() []*sqlf.Query {
 	conds := []*sqlf.Query{sqlf.Sprintf("deleted_at IS NULL")}
 	if len(o.Kinds) > 0 {
-		kinds := []*sqlf.Query{}
+		kinds := make([]*sqlf.Query, 0, len(o.Kinds))
 		for _, kind := range o.Kinds {
 			kinds = append(kinds, sqlf.Sprintf("%s", kind))
 		}
-		conds = append(conds, sqlf.Sprintf("kind IN (%s)", sqlf.Join(kinds, ", ")))
+		conds = append(conds, sqlf.Sprintf("kind IN (%s)", sqlf.Join(kinds, ",")))
```
Comment (unknwon) [inline cmd/frontend/db/external_services.go:69]: `sqlf.Join` always adds one space to both sides of this argument (`sep`).

File: cmd/frontend/db/external_services.go [hunk 2]
```diff
@@ -272,7 +272,7 @@ func (e *ExternalServicesStore) Create(ctx context.Context, confGet func() *conf
 		return err
 	}
 
-	externalService.CreatedAt = time.Now()
+	externalService.CreatedAt = time.Now().UTC().Truncate(time.Microsecond)
```
Comment (unknwon) [inline cmd/frontend/db/external_services.go:275]: Postgres 9.6 does not support microsecond. Truncate it so make tests happy, and align with what we do in other places: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+Truncate%28time.Microsecond%29&patternType=literal

File: cmd/frontend/db/external_services.go [hunk 3]
```diff
@@ -372,13 +372,16 @@ func (e *ExternalServicesStore) GetByID(ctx context.Context, id int64) (*types.E
 		return Mocks.ExternalServices.GetByID(id)
 	}
 
-	conds := []*sqlf.Query{sqlf.Sprintf("id=%d", id)}
+	conds := []*sqlf.Query{
+		sqlf.Sprintf("deleted_at IS NULL"),
```
Comment (unknwon) [inline cmd/frontend/db/external_services.go:376]: We do soft-delete for external services, and filter out soft-deleted records in other queries, so I think this was an oversight.

File: cmd/frontend/db/external_services_test.go [hunk 4]
```diff
@@ -59,20 +109,236 @@ func TestExternalServicesStore_ValidateConfig(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if len(ess) != 1 {
+		t.Fatalf("Want 1 external service but got %d", len(ess))
+	} else if diff := cmp.Diff(es, ess[0]); diff != "" {
+		t.Fatalf("(-want +got):\n%s", diff)
+	}
+}
+func TestExternalServicesStore_Count(t *testing.T) {
```
Comment (ryanslade) [inline cmd/frontend/db/external_services_test.go:315]: Nititiest of nitpicks: Missing space after previous function 😄

File: cmd/frontend/db/external_services_test.go [hunk 5]
```diff
@@ -2,43 +2,93 @@ package db
...
 	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/keegancsmith/sqlf"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbtesting"
 )
 
+func TestExternalServicesListOptions_sqlConditions(t *testing.T) {
```
Comment (tsenart) [inline cmd/frontend/db/external_services_test.go:15]: Not sure it's worth unit testing this if we are already doing integration tests for the methods that call this underneath. What do you think?

File: cmd/frontend/db/external_services.go (hunk 2 above)
Comment (unknwon) [reply cmd/frontend/db/external_services.go:275]: Filed in https://github.com/sourcegraph/sourcegraph/issues/11246

File: cmd/frontend/db/external_services_test.go (hunk 4 above)
Comment (unknwon) [reply cmd/frontend/db/external_services_test.go:315]: Ah... I did partial stage and missed the line break!

File: cmd/frontend/db/external_services.go [hunk 6]
```diff
@@ -367,37 +367,40 @@ func (*ExternalServicesStore) Delete(ctx context.Context, id int64) error {
...
 		return Mocks.ExternalServices.GetByID(id)
 	}
 
-	conds := []*sqlf.Query{sqlf.Sprintf("id=%d", id)}
-	ExternalServicesStore, err := c.list(ctx, conds, nil)
+	conds := []*sqlf.Query{
+		sqlf.Sprintf("deleted_at IS NULL"),
+		sqlf.Sprintf("id=%d", id),
+	}
+	ExternalServicesStore, err := e.list(ctx, conds, nil)
```
[the remark this answers is not in the document]
Comment (unknwon) [reply cmd/frontend/db/external_services.go]: >Seems like a search and replace #fail to me

LOL, good catch, didn't even notice.

File: cmd/frontend/db/external_services_test.go (hunk 5 above)
Comment (unknwon) [reply cmd/frontend/db/external_services_test.go:15]: Hmm... I would like to keep it for sanity check.

**Pull request merged**